pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,242 @@
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer: cache.NewIndexer(keyFunc, cache.Indexers{
+					common.TagScopeSubnetCRUID: subnetIndexFunc,
+				}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, nil, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+// TODO Test update of VpcSubnet(eg. update tags)
+func (service *SubnetService) CreateOrUpdateSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo, tags []model.Tag) (string, error) {
+	value, ok := obj.GetAnnotations()[common.TagScopeSubnetType]
+	if ok {
+		tags = append(tags, model.Tag{
+			Scope: String(common.TagScopeSubnetType),
+			Tag:   String(value),
+		})
+	}
+	uid := string(obj.GetUID())
+	nsxSubnet, err := service.buildSubnet(obj, tags)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return "", err
+	}
+	switch obj.(type) {
+	case *v1alpha1.Subnet:
+		//TODO Additional check for len(nsxSubnets)==1 for improvement.
+		nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, uid)
+		changed := false
+		if len(nsxSubnets) == 0 {
+			changed = true
+		} else {
+			changed = common.CompareResource(SubnetToComparable(&nsxSubnets[0]), SubnetToComparable(nsxSubnet))
+		}
+		if !changed {
+			log.Info("subnet not changed, skip updating", "subnet.Id", *nsxSubnets[0].Id)
+			return *nsxSubnets[0].Id, nil
+		}
+	}
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, vpcInfo)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return "", err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return "", err
+	}
+	realizeService := realizestate.InitializeRealizeState(service.Service)
+	if err = realizeService.CheckRealizeState(retry.DefaultRetry, *nsxSubnet.Path, "RealizedLogicalSwitch"); err != nil {
+		return "", err
+	}
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSubnet.Id); err != nil {
+		return "", err
+	}
+	if err = service.SubnetStore.Operate(nsxSubnet); err != nil {
+		return "", err
+	}
+	log.Info("successfully updated nsxSubnet", "nsxSubnet", nsxSubnet)
+	return *nsxSubnet.Id, nil
+}
+
+func (service *SubnetService) DeleteSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo) error {
+	nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, string(obj.GetUID()))
+	for _, nsxSubnet := range nsxSubnets {
+		nsxSubnet.MarkedForDelete = &MarkedForDelete
```

Comment:
is it possible that we check if there is subnet ports on the subnet before invoking NSX API to delete the subnet?
If still subnet ports on it, re-enqueue the event.

Reply:
It will require another api call for subnet ports in subnet service, so I would prefer to check if there are subnet ports left from subnet store as we can regard store as the source of truth. How do you think?